File: system-test/dotnet/Dsl.Core/Shared/UseCaseResult.cs
using Common;

namespace Dsl.Core.Shared;

public class UseCaseResult<TSuccessResponse, TFailureResponse, TSuccessVerification, TFailureVerification>
{
    private readonly Result<TSuccessResponse, TFailureResponse> _result;
    private readonly UseCaseContext _context;
    private readonly Func<TSuccessResponse, UseCaseContext, TSuccessVerification> _successVerificationFactory;
    private readonly Func<TFailureResponse, UseCaseContext, TFailureVerification> _failureVerificationFactory;

    public UseCaseResult(
        Result<TSuccessResponse, TFailureResponse> result,
        UseCaseContext context,
        Func<TSuccessResponse, UseCaseContext, TSuccessVerification> successVerificationFactory,
        Func<TFailureResponse, UseCaseContext, TFailureVerification> failureVerificationFactory)
    {
        _result = result;
        _context = context;
        _successVerificationFactory = successVerificationFactory;
        _failureVerificationFactory = failureVerificationFactory;
    }

    public TSuccessVerification ShouldSucceed()
    {
        _result.ShouldBeSuccess();
        return _successVerificationFactory(_result.Value, _context);
    }

    public TFailureVerification ShouldFail()
    {
        _result.ShouldBeFailure();
        return _failureVerificationFactory(_result.Error, _context);
    }
}

File: system-test/java/src/main/java/com/optivem/shop/dsl/core/shared/UseCaseResult.java
package com.optivem.shop.dsl.core.shared;

import com.optivem.shop.dsl.driver.port.shared.dtos.ErrorResponse;
import com.optivem.shop.dsl.common.Result;

import java.util.function.BiFunction;

import static com.optivem.shop.dsl.common.ResultAssert.assertThatResult;

public class UseCaseResult<TSuccessResponse, TSuccessVerification> {
    private final Result<TSuccessResponse, ErrorResponse> result;
    private final UseCaseContext context;
    private final BiFunction<TSuccessResponse, UseCaseContext, TSuccessVerification> successVerificationFactory;

    public UseCaseResult(
            Result<TSuccessResponse, ErrorResponse> result,
            UseCaseContext context,
            BiFunction<TSuccessResponse, UseCaseContext, TSuccessVerification> successVerificationFactory) {
        this.result = result;
        this.context = context;
        this.successVerificationFactory = successVerificationFactory;
    }

    public TSuccessVerification shouldSucceed() {
        assertThatResult(result).isSuccess();
        return successVerificationFactory.apply(result.getValue(), context);
    }

    public ErrorVerification shouldFail() {
        assertThatResult(result).isFailure();
        return new ErrorVerification(result.getError(), context);
    }
}




